### PR TITLE
Add sequence alterations and other feature types to GFF3+FASTA export

### LINKED ISSUE #748: Sequence modifications aren't included in gff3 export
Hi again,

I've created some sequence mods (insertion, substitution, deletion) in the Apollo 2 demo instance:
http://icebox.lbl.gov/Apollo2/annotator/loadLink?loc=Group1.1:725082..725173&organism=3836&tracks=Official Gene Set v3.2
In Web Apollo 1, the modifications were retrievable from the exported gff3 file. However, they weren't present in the exported gff3 file from this instance (I used the export options from the 'Ref Sequence' tab). 

Apologies if this is due to my lack of familiarity with Apollo 2!

### PR DESCRIPTION
**Pull request by cmdcolin**
Fixes #748 and #784, because the gff3 exporter service was hardcoded to only export gene types. No reason though. Works fine to export other things including sequence alterations.

Was fixed awhile ago but I was not used to this whole fork thing....so here it is finally.

### CODE REVIEW

**Comment by nathandunn** [comment]
Thanks.  I'll take a look.

File: grails-app/controllers/org/bbop/apollo/IOServiceController.groovy [hunk 1]
```diff
@@ -93,7 +94,7 @@ class IOServiceController extends AbstractApolloController {
                         }
                     }
                 }
-                'in'('class',Gene.class.name)
+                'in'('class',requestHandlingService.viewableAnnotationList+[Deletion.class.name,Insertion.class.name,Substitution.class.name])
```
Comment (nathandunn) [inline grails-app/controllers/org/bbop/apollo/IOServiceController.groovy]: This is fine for now . . probably good to have somewhere: viewableAlterationList . . . . but there are other refactoring issues with this, so good for now.

**Comment by nathandunn** [comment]
I think I had another comment which should be implemented, but I am also getting this error when exporting FASTA:

```
| Error groovy.lang.MissingMethodException: No signature of method: org.bbop.apollo.TranscriptService.getTranscripts() is applicable for argument types: (org.bbop.apollo.Insertion) values: [Feature{id=63, symbol='null', description='null', name='79dc0a73-6910-447e-909a-bb61c3328f02-insertion', uniqueName='79dc0a73-6910-447e-909a-bb61c3328f02', sequenceLength=null, status=null, dateCreated=2016-01-22 17:02:51.291, lastUpdated=2016-01-22 17:02:51.325}]
Possible solutions: getTranscripts(org.bbop.apollo.Gene), getTranscript(org.bbop.apollo.CDS)
```

**Comment by nathandunn** [comment]
Probably a good reason to have some integration tests to add sequence alterations, which would catch stuff like this.    I think that GFF3HandlerServiceIntegrationSpec has a good starting point for this.

**Comment by nathandunn** [comment]
But the GFF3 + FASTA is all there it looks like:

Looks like its all there for GFF3 + FASTA . .  

```
###
chrI	.	pseudogene	192613	196179	.	+	.	Name=YAR042Wa;date_creation=2016-01-22;owner=ndunn@me.com;ID=e9bd3a27-9e30-4b14-b2b2-f680ecfaa8f7;date_last_modified=2016-01-22
chrI	.	Transcript	192613	196179	.	+	.	Name=YAR042Wa-00001;date_creation=2016-01-22;Parent=e9bd3a27-9e30-4b14-b2b2-f680ecfaa8f7;owner=ndunn@me.com;ID=bb8bad45-512a-4c72-bf6c-62c584e249e6;date_last_modified=2016-01-22
chrI	.	exon	192613	196179	.	+	.	Name=518c3d66-14c8-4c03-a64b-4c95954c0f5d-exon;date_creation=2016-01-22;Parent=bb8bad45-512a-4c72-bf6c-62c584e249e6;ID=518c3d66-14c8-4c03-a64b-4c95954c0f5d;date_last_modified=2016-01-22
###
chrI	.	insertion	193916	193916	.	+	.	Name=79dc0a73-6910-447e-909a-bb61c3328f02-insertion;date_creation=2016-01-22;owner=ndunn@me.com;ID=79dc0a73-6910-447e-909a-bb61c3328f02;justification=genome insertion comment;date_last_modified=2016-01-22
###
chrI	.	deletion	193925	193930	.	+	.	Name=e6322341-9a6e-435a-9d47-46849c562160-deletion;date_creation=2016-01-22;owner=ndunn@me.com;ID=e6322341-9a6e-435a-9d47-46849c562160;justification=deletion comment;date_last_modified=2016-01-22
###
chrI	.	substitution	193937	193948	.	+	.	Name=c797930a-6e78-4c78-afed-e2c14bddc166-substitution;date_creation=2016-01-22;owner=ndunn@me.com;ID=c797930a-6e78-4c78-afed-e2c14bddc166;justification=sub comment;date_last_modified=2016-01-22
###
chrI	.	insertion	204829	204829	.	+	.	Name=534fda93-bea7-41b9-aadf-4a2b93607427-insertion;date_creation=2016-01-22;owner=ndunn@me.com;ID=534fda93-bea7-41b9-aadf-4a2b93607427;justification=insertion comment;date_last_modified=2016-01-22
###
chrI	.	deletion	204844	204850	.	+	.	Name=d5a29a7d-1206-4f8f-ad9b-165c847d4efb-deletion;date_creation=2016-01-22;owner=ndunn@me.com;ID=d5a29a7d-1206-4f8f-ad9b-165c847d4efb;justification=deletion comment;date_last_modified=2016-01-22
###
chrI	.	substitution	204858	204870	.	+	.	Name=f26a9d41-5d27-4fac-af3c-8912090e0a97-substitution;date_creation=2016-01-22;owner=ndunn@me.com;ID=f26a9d41-5d27-4fac-af3c-8912090e0a97;justification=sub comment 2;date_last_modified=2016-01-22
###
```

with included FASTA:
```
>79dc0a73-6910-447e-909a-bb61c3328f02-insertion
AAAA
>e6322341-9a6e-435a-9d47-46849c562160-deletion
ATGACG
>c797930a-6e78-4c78-afed-e2c14bddc166-substitution
ATATAGAGAGAT
>534fda93-bea7-41b9-aadf-4a2b93607427-insertion
AAATATAGA
>d5a29a7d-1206-4f8f-ad9b-165c847d4efb-deletion
CAACTGA
>f26a9d41-5d27-4fac-af3c-8912090e0a97-substitution
AATAGAGAGATAC
```

**Comment by nathandunn** [comment]
@cmdcolin   See comment:  

https://github.com/GMOD/Apollo/pull/822#issuecomment-174102515

Also, please add some sort of integration test, if for at least to cover the basics.  Group1.10 should be in the integration tests with a lot of examples.   Please ask @deepakunni3  if you need help writing these.

**Comment by cmdcolin** [comment]
I guess short of a full gff3 parser here, I might just add some string similarity checks to the integration test

**Comment by nathandunn** [comment]
I think that would be fine.   I would look for keywords and then add very "unique" sequence alterations to test with to verify if they are there or not.    Thx

**Comment by cmdcolin** [comment]
@nathandunn  added a basic Gff3HandlerService integration test that just checks for existence of a gene and sequence alteration in gff3 output

**Comment by nathandunn** [comment]
I think that this test will be good enough for GFF3.  Can we do something similar for FASTA?

File: grails-app/conf/Config.groovy [hunk 2]
```diff
@@ -167,6 +167,7 @@ log4j.main = {
 	warn 'grails.app'
 //    debug 'grails.app'
 //    debug 'liquibase'
+//    debug 'org.bbop.apollo'
```
Comment (nathandunn) [inline grails-app/conf/Config.groovy]: can probably remove, unless you want to leave it for reference.  I think you want:

```debug 'grails.app````

File: grails-app/services/org/bbop/apollo/Gff3HandlerService.groovy [hunk 3]
```diff
@@ -304,8 +305,8 @@ public class Gff3HandlerService {
                 attributes.put(FeatureStringEnum.EXPORT_ALIAS.value, synonyms.toString());
             }
         }
-
-        if (feature.ontologyId != Gene.ontologyId) {
+        if (!(requestHandlingService.viewableAnnotationList+[Deletion.class.name,Insertion.class.name,Substitution.class.name]).contains(feature.class.name)) {
+        //if (!(feature.ontologyId in (requestHandlingService.viewableAnnotationList+[Deletion.class.name,Insertion.class.name,Substitution.class.name]))) {
```
Comment (nathandunn) [inline grails-app/services/org/bbop/apollo/Gff3HandlerService.groovy]: can probably remove this line

File: grails-app/conf/Config.groovy (hunk 2 above)
Comment (cmdcolin) [inline grails-app/conf/Config.groovy]: adding this allows debug logging in integration tests. as far as i can tell, didnt output debug logging with just grails.app

Comment (nathandunn) [inline grails-app/conf/Config.groovy]: sounds good . . didn't think of that.

Comment (nathandunn) [inline grails-app/conf/Config.groovy]: :+1:

**Comment by cmdcolin** [comment]
@deepakunni3 and I noticed that the Gff3HandlerServiceIntegrationSpec was probably made pretty early in development and has some weird assumptions, namely that it makes multiple CDS per transcript

I'm recoding it to just use a more straightforward methodology and adding the Fasta version

**Comment by cmdcolin** [comment]
@nathandunn added a simple FASTA integration test and cleaned up the Gff3 one

**Comment by nathandunn** [comment]
Thanks for cleaning of the GFF3 one.  Yes, definitely older tests.  Testing momentarily.

**Comment by nathandunn** [comment]
I'm getting this error when trying to export FASTA (Genomic) from Group1.10 with two annotations (when MRNA and one Pseudogene):

```
2016-01-29 11:29:50,237 [http-bio-8080-exec-13] WARN  apollo.JbrowseController  - File not found: /opt/apollo/honeybee/data//tracks.conf
| Error groovy.lang.MissingMethodException: No signature of method: org.bbop.apollo.TranscriptService.getTranscripts() is applicable for argument types: (org.bbop.apollo.Insertion) values: [Feature{id=196609, symbol='null', description='null', name='1b38ce7b-f5bf-474a-b1fc-e439a2cbb305-insertion', uniqueName='1b38ce7b-f5bf-474a-b1fc-e439a2cbb305', sequenceLength=null, status=null, dateCreated=2016-01-29 11:27:58.757, lastUpdated=2016-01-29 11:27:58.771}]
Possible solutions: getTranscripts(org.bbop.apollo.Gene), getTranscript(org.bbop.apollo.CDS)
| Error 	at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.unwrap(ScriptBytecodeAdapter.java:56)
| Error 	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:51)
| Error 	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
| Error 	at org.bbop.apollo.FastaHandlerService.writeFeatures(FastaHandlerService.groovy:111)
| Error 	at org.bbop.apollo.FastaHandlerService$writeFeatures.callCurrent(Unknown Source)
| Error 	at org.bbop.apollo.FastaHandlerService.writeFeatures(FastaHandlerService.groovy:100)
| Error 	at org.bbop.apollo.FastaHandlerService$$FastClassBySpringCGLIB$$ed339c14.invoke(<generated>)
| Error 	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
| Error 	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:708)
| Error 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
| Error 	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:98)
| Error 	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:262)
| Error 	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:95)
| Error 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
| Error 	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:644)
| Error 	at org.bbop.apollo.FastaHandlerService$$EnhancerBySpringCGLIB$$14d2bc4f.writeFeatures(<generated>)
| Error 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
| Error 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
| Error 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
| Error 	at java.lang.reflect.Method.invoke(Method.java:606)
| Error 	at org.springsource.loaded.ri.ReflectiveInterceptor.jlrMethodInvoke(ReflectiveInterceptor.java:1270)
| Error 	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSite.invoke(PogoMetaMethodSite.java:166)
| Error 	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.call(PogoMetaMethodSite.java:68)
| Error 	at org.bbop.apollo.IOServiceController._do_write(IOServiceController.groovy:131)
| Error 	at org.bbop.apollo.IOServiceController.this$3$_do_write(IOServiceController.groovy)
| Error 	at org.bbop.apollo.IOServiceController$this$3$_do_write.callCurrent(Unknown Source)
| Error 	at org.bbop.apollo.IOServiceController.write(IOServiceController.groovy)
| Error 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
| Error 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
| Error 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
| Error 	at java.lang.reflect.Method.invoke(Method.java:606)
| Error 	at org.springsource.loaded.ri.ReflectiveInterceptor.jlrMethodInvoke(ReflectiveInterceptor.java:12
```

**Comment by cmdcolin** [comment]
Interesting case. One way to fix is to check that it is a gene/pseudogene before doing the getTranscripts call in FastaHandlerService (previously it was trying to call getTranscripts on every feature that came into the fasta service, but if we have other features, it should avoid that)

5e034a1 may be sufficient to patch

**Comment by nathandunn** [comment]
Can you test and let me know when its ready.  Thanks.

**Comment by cmdcolin** [comment]
I think it should be ready. I tested with various combinations (fasta/gff with insertions/deletions, repeat region, pseudogene, gene, all scaffolds, etc.) and added a pseudogene to the gff3 test.

**Comment by nathandunn** [comment]
thanks . . I'll give it another shot

**Comment by nathandunn** [comment]
:+1:

**Pull request merged**